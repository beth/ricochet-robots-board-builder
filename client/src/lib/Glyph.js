import COLORS from './config/colors';
import TOKEN_DESIGNS from './config/tokenDesigns';

class Glyph {
  constructor(color, design) {
    this.color = color;
    this.design = design;
  }

  getName() {
    return `${this.design}-${this.color}`;
  }
}

const generateGlyphs = () => {
  const glyphs = [];
  COLORS.forEach((color) => {
    TOKEN_DESIGNS.forEach((design) => {
      glyphs.push(new Glyph(color, design));
    });
  });
  glyphs.push(new Glyph('ALL', 'ALL'));
  return glyphs;
};

export {
  Glyph,
  generateGlyphs,
};
