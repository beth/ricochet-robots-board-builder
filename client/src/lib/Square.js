import SIZE from './config/size';

const allowedWalls = new Set(['top', 'bottom', 'left', 'right']);

class Square {
  constructor(row, col) {
    this.row = row;
    this.col = col;
    this.walls = {};
    this.glyph = null;
    this.connector = null;
  }

  setWall(wall) {
    this.walls[wall] = true;
  }

  removeWall(wall) {
    delete this.walls[wall];
  }

  modifyWall(wall, val) {
    if (val !== undefined && allowedWalls.has(wall)) {
      if (val) {
        this.setWall(wall);
      } else {
        this.removeWall(wall);
      }
    }
  }

  modifyWalls(wallsToModify) {
    const walls = Object.entries(wallsToModify);
    const WALL = 0;
    const VAL = 1;
    for (let i = 0; i < walls.length; i += 1) {
      this.modifyWall(walls[i][WALL], walls[i][VAL]);
    }
  }

  setGlyph(name) {
    this.glyph = name;
  }

  removeGlyph() {
    this.glyph = null;
  }

  setConnector(color) {
    this.connector = color;
  }
}

const addWalls = (squares, [rStart, rStop = rStart], [cStart, cStop = cStart], wall) => {
  for (let row = rStart; row <= rStop; row += 1) {
    for (let col = cStart; col <= cStop; col += 1) {
      squares[row][col].setWall(wall);
    }
  }
};

const addBorderWalls = (squares) => {
  addWalls(squares, [0], [0, (SIZE * 2) - 1], 'top');
  addWalls(squares, [(SIZE * 2) - 1], [0, (SIZE * 2) - 1], 'bottom');
  addWalls(squares, [0, (SIZE * 2) - 1], [0], 'left');
  addWalls(squares, [0, (SIZE * 2) - 1], [(SIZE * 2) - 1], 'right');
};

const addConnectorSquares = (squares) => {
  squares[SIZE - 1][SIZE - 1].modifyWalls({ top: true, left: true });
  squares[SIZE - 1][SIZE].modifyWalls({ top: true, right: true });
  squares[SIZE][SIZE - 1].modifyWalls({ bottom: true, left: true });
  squares[SIZE][SIZE].modifyWalls({ bottom: true, right: true });
  squares[SIZE - 1][SIZE - 1].setConnector(true);
  squares[SIZE - 1][SIZE].setConnector(true);
  squares[SIZE][SIZE - 1].setConnector(true);
  squares[SIZE][SIZE].setConnector(true);
};

const generateBlankSquares = () => {
  const squares = [];
  for (let row = 0; row < SIZE * 2; row += 1) {
    squares.push([]);
    for (let col = 0; col < SIZE * 2; col += 1) {
      squares[row].push(new Square(row, col));
    }
  }
  return squares;
};

const generateSquares = () => {
  const squares = generateBlankSquares();
  addBorderWalls(squares);
  addConnectorSquares(squares);
  return squares;
};

export {
  Square,
  generateSquares,
};

