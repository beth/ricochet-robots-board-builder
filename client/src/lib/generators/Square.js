const Square = () => ({
  walls: {},
  glyph: null,
  connector: null,
});

const generateBlankSquares = (size) => {
  const squares = [];
  for (let row = 0; row < size * 2; row += 1) {
    squares.push([]);
    for (let col = 0; col < size * 2; col += 1) {
      squares[row].push(new Square());
    }
  }
  return squares;
};

const generateSquares = (size) => {
  const squares = generateBlankSquares(size);

  const addWalls = ([rStart, rStop = rStart], [cStart, cStop = cStart], wall) => {
    for (let row = rStart; row <= rStop; row += 1) {
      for (let col = cStart; col <= cStop; col += 1) {
        squares[row][col].walls[wall] = true;
      }
    }
  };

  const addBorderWalls = () => {
    addWalls([0], [0, (size * 2) - 1], 'top');
    addWalls([(size * 2) - 1], [0, (size * 2) - 1], 'bottom');
    addWalls([0, (size * 2) - 1], [0], 'left');
    addWalls([0, (size * 2) - 1], [(size * 2) - 1], 'right');
  };

  const addConnectorWalls = () => {
    addWalls([size - 1], [size - 1, size], 'top');
    addWalls([size], [size - 1, size], 'bottom');
    addWalls([size - 1, size], [size - 1], 'left');
    addWalls([size - 1, size], [size], 'right');
  };

  const addConnectorSquares = () => {
    squares[size - 1][size - 1].connector = true;
    squares[size - 1][size].connector = true;
    squares[size][size - 1].connector = true;
    squares[size][size].connector = true;
  };

  addBorderWalls(size);
  addConnectorSquares(size);
  addConnectorWalls(size);
  return squares;
};

export default generateSquares;
