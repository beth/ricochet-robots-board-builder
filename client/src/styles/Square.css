:local .square {
  border-style: solid;
  border-color: #dfdfdf;
  border-width: 0.5px;
  display: grid;
  align-content: space-around;
  justify-content: center;
}

:local .connector {
  background: #dfdfdf;
}

:local .top {
  border-top-color: black;
  border-top-width: 4px;
  border-top-style: solid;
}

:local .bottom {
  border-bottom-color: black;
  border-bottom-width: 4px;
  border-bottom-style: solid;
}

:local .left {
  border-left-color: black;
  border-left-width: 4px;
  border-left-style: solid;
}

:local .right {
  border-right-color: black;
  border-right-width: 4px;
  border-right-style: solid;
}
