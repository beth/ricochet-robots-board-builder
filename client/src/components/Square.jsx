import React from 'react';
import PropTypes from 'prop-types';
import styles from '../styles/Square.css';

const classes = (square) => {
  const classNames = Object.keys(square.walls).map(wall => styles[wall]);
  classNames.push(styles.square);
  if (square.connector) {
    classNames.push(styles.connector);
  }
  return classNames.join(' ');
};

const Square = props => (
  <div className={classes(props.square)} />
);

Square.propTypes = {
  square: PropTypes.shape({ walls: PropTypes.object }).isRequired,
};

export default Square;
