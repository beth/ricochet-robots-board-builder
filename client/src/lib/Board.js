import GLYPH_QUADRANTS from './config/glyphQuadrants';
import * as square from './Square';

/*
  BOARD HELPERS
*/
const determineQuadrant = (size, row, col) => {
  if (row <= size - 1 && col <= size - 1) {
    return 'TL';
  } else if (row <= size - 1) {
    return 'TR';
  } else if (col <= size - 1) {
    return 'BL';
  }
  return 'BR';
};

const glyphAllowedInColorQuadrant = (glyph, color) => GLYPH_QUADRANTS[glyph.getName()] === color;

const isConnector = (board, row, col) => Boolean(board.getIn(['squares', row, col, 'connector']));

const remainingQuadrant = board => board.get('quadrants').reduce((memo, color, quadrant) => (color ? memo : quadrant), null);

/*
  BOARD METHODS
*/
const setQuadrant = (board, quadrant, color) => {
  let newBoard = board;
  if (!board.get('remainingColors').has(color)) {
    return board;
  }
  const previousColor = board.getIn(['quadrants', quadrant]);
  if (previousColor) {
    newBoard = newBoard.update('remainingColors', remainingColors => remainingColors.add(previousColor));
  }
  newBoard = newBoard.setIn(['quadrants', quadrant], color);
  newBoard = newBoard.update('remainingColors', remainingColors => remainingColors.delete(color));
  if (newBoard.get('remainingColors').size === 1) {
    newBoard = setQuadrant(newBoard, remainingQuadrant(newBoard), newBoard.get('remainingColors').first());
  }
  return newBoard;
};

const setGlyph = (board, glyph, row, col) => {
  let newBoard = board;
  const quadrant = determineQuadrant(board.get('size'), row, col);
  if (!glyphAllowedInColorQuadrant(glyph, board.getIn(['quadrants', quadrant])) || isConnector(board, row, col)) {
    return newBoard;
  }
  const currentGlyph = board.getIn(['squares', row, col, 'glyph']);
  if (currentGlyph) {
    newBoard = newBoard.update('remainingGlyphs', remainingGlyphs => remainingGlyphs.add(currentGlyph));
  }
  newBoard = newBoard.updateIn(['squares', row, col], squareToUpdate => square.setGlyph(squareToUpdate, glyph));
  newBoard = newBoard.update('remainingGlyphs', remainingGlyphs => remainingGlyphs.delete(glyph));
  return newBoard;
};

export {
  determineQuadrant,
  glyphAllowedInColorQuadrant,
  isConnector,
  remainingQuadrant,
  setQuadrant,
  setGlyph,
};
