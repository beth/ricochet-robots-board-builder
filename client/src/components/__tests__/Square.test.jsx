import React from 'react';
import renderer from 'react-test-renderer';
import Square from '../Square';
import { Square as SquareConstructor } from '../../lib/Square';

describe('Square Component', () => {
  it('renders correctly', () => {
    const square = new SquareConstructor(0, 0);
    const domTree = renderer.create(<Square square={square} />).toJSON();
    expect(domTree).toMatchSnapshot();
  });

  it('renders correctly when walls are present', () => {
    const square = new SquareConstructor(0, 0);
    square.modifyWalls({ top: true, bottom: true });
    const domTree = renderer.create(<Square square={square} />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
