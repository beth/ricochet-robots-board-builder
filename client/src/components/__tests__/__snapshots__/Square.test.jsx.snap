// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Square Component renders correctly 1`] = `
<div
  className="square"
/>
`;

exports[`Square Component renders correctly when walls are present 1`] = `
<div
  className="top bottom square"
/>
`;
